File: Others/hard/TopologicalSort.py
def make_graph(jobs, deps):
    graph = {id: [] for id in jobs}

    for dep in deps:
        graph[dep[0]].append(dep[1])
    return graph


def topologicalSort(jobs, deps):
    in_degree = {id: 0 for id in jobs}

    for dep in deps:
        in_degree[dep[1]] += 1

    queue = [x for x in in_degree if in_degree[x] == 0]
    result = []
    graph = make_graph(jobs, deps)

    while queue:
        current_node = queue.pop(0)

        result.append(current_node)

        for adj in graph[current_node]:
            in_degree[adj] -= 1
            if in_degree[adj] == 0:
                queue.append(adj)

    return result if len(result) == len(jobs) else []

File: Golang/hard/topological-sort.go
package main

type Dep struct {
	Prereq int
	Job    int
}

type Graph map[int][]int

type Queue []int

func (q *Queue) Push(elem int) {
	*q = append(*q, elem)
}

func (q *Queue) Dequeue() int {
	if q.IsEmpty() {
		panic("Dequeuing from an empty queue...")
	}
	toRet := (*q)[0]
	*q = (*q)[1:]
	return toRet
}

func (q Queue) IsEmpty() bool {
	return len(q) == 0
}

func TopologicalSort(jobs []int, deps []Dep) []int {
	graph := Graph{}
	inDeg := map[int]int{}

	for _, job := range jobs {
		inDeg[job] = 0
		graph[job] = []int{}
	}

	for _, dep := range deps {
		inDeg[dep.Job]++
		graph[dep.Prereq] = append(graph[dep.Prereq], dep.Job)
	}

	// TOPO SORT
	queue := Queue{}

	for job, deg := range inDeg {
		if deg == 0 {
			queue.Push(job)
		}
	}

	res := []int{}

	for !queue.IsEmpty() {
		node := queue.Dequeue()
		res = append(res, node)
		for _, adj := range graph[node] {
			inDeg[adj]--
			if inDeg[adj] == 0 {
				queue = append(queue, adj)
			}
		}
	}

	if len(res) != len(jobs) {
		return []int{}
	}
	return res
}
